### PR TITLE
Child base contract 

### PR DESCRIPTION
**Pull request by 0xAshish**

### CODE REVIEW

File: contracts/child/ChildERC20.sol [hunk 1]
```diff
@@ -18,13 +19,19 @@ contract ChildERC20 is ChildToken, ERC20, ERC20Detailed {
     uint256 output2
   );
   // constructor
-  constructor (address _token, string _name, string _symbol, uint8 _decimals)
+  constructor (address _owner, address _token, string _name, string _symbol, uint8 _decimals)
     public
     ERC20Detailed(_name, _symbol, _decimals) {
-    require(_token != address(0));
+    require(_token != address(0x0) && _owner != address(0x0));
+    parentOwner = _owner;
```
Comment (jdkanani) [inline contracts/child/ChildERC20.sol]: `parent` or `parentOwner`

File: contracts/child/ChildERC721.sol [hunk 2]
```diff
@@ -15,14 +16,19 @@ contract ChildERC721 is ChildToken, ERC721Full {
...
 
   // constructor
-  constructor (address _token, string name, string symbol) ERC721Full(name, symbol)
+  constructor (address _owner, address _token, string name, string symbol) ERC721Full(name, symbol)
     public 
     {
-    require(_token != address(0));
-
+    require(_token != address(0x0) && _owner != address(0x0));
+    parentOwner = _owner;
```
Comment (jdkanani) [inline contracts/child/ChildERC721.sol]: `parentOwner` or `parent`?

File: contracts/child/IParentToken.sol [hunk 3]
```diff
@@ -0,0 +1,7 @@
+pragma solidity ^0.4.24;
+//interface for parent contract of any child token
+
+
+interface IParentToken {
+  function beforeTransfer(address user) external returns(bool);
+}
```
Comment (jdkanani) [inline contracts/child/IParentToken.sol]: Add empty end-line

File: contracts/child/ParentToken.sol [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  mapping (address => bool) isAllowed;
+  function beforeTransfer(address user) public returns(bool) {
+    return isAllowed[user];
+  }
+
+  function updatePermission(address user) public onlyOwner {
+    require(user != address(0x0));
+    isAllowed[user] = !isAllowed[user];
+  }
+}
```
Comment (jdkanani) [inline contracts/child/ParentToken.sol]: Add end empty line

File: contracts/child/ChildERC721.sol (hunk 2 above)
Comment (0xAshish) [reply contracts/child/ChildERC721.sol]: `parentOwner` is an owner of parent Contract.
`parent` is the address of parent contract.
show we rename ?

**Comment by jdkanani** [comment]
@0xAshish Please fix ci

File: contracts/child/ChildERC20.sol [hunk 5]
```diff
@@ -69,7 +76,10 @@ contract ChildERC20 is ChildToken, ERC20, ERC20Detailed {
   /// @param to Address of token receiver.
   /// @param value Number of tokens to transfer.
   /// @return Returns success of function call.
-  function transfer( address to, uint256 value) public returns (bool) {
+  function transfer(address to, uint256 value) public returns (bool) {
+    if (parent != address(0x0) && !IParentToken(parent).beforeTransfer(msg.sender)) {
```
Comment (jdkanani) [inline contracts/child/ChildERC20.sol]: Can we change to `beforeTransfer(address sender, address to, uint256 value)`

File: contracts/child/IParentToken.sol [hunk 6]
```diff
@@ -0,0 +1,7 @@
+pragma solidity ^0.4.24;
+//interface for parent contract of any child token
+
+
+interface IParentToken {
+  function beforeTransfer(address user) public returns(bool);
```
Comment (jdkanani) [inline contracts/child/IParentToken.sol]: Please add two more arguments -- `(address to, uint256 value)`

File: contracts/child/ParentToken.sol [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+// demo token parent contract  
+
+
+contract ParentToken is IParentToken, Ownable {
+  mapping (address => bool) isAllowed;
+  function beforeTransfer(address user) public returns(bool) {
+    return isAllowed[user];
+  }
+
+  function updatePermission(address user) public onlyOwner {
```
Comment (jdkanani) [inline contracts/child/ParentToken.sol]: Change it to `ParentTokenMock`

File: contracts/child/IParentToken.sol (hunk 6 above)
Comment (0xAshish) [reply contracts/child/IParentToken.sol]: added

**Pull request merged**